### PR TITLE
Feat/rpc comms layer

### PR DESCRIPTION
**Pull request by petioptrv**
**Before submitting this PR, please make sure**:

- [X] Your code builds clean without any errors or warnings
- [X] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:
This PR lays the foundation for a layer of external communication with the bot. 

A new class `HummingbotController` is introduced that currently does not contain any logic, but once we start exposing more complex commands like `config_command`, then this class will handle the additional logic.

On top of that class is introduced `HummingbotControllerRPCServer`, which exposes `HummingbotController`'s method via gRPC.

The `hummingbot_quickstart.py` entry script has been adapted to accept a new flag `-c` that will start the RPC server.

Finally, a new script has been introduced `bot_controller_client_demo.py`, that allows quick testing of the server.

**Tests performed by the developer**:
- Tested the server on PMM, stop, start, exit.
- Tested running quickstart without the server.
- Basic unit tests.

**Tips for QA testing**:
Ensure that the bot can be controlled via the RPC server by using the demo controller.

### CODE REVIEW

**Comment by Deepali6997** [comment]
Can confirm that it is working fine for source version -- Also, as confirmed by @gerald-hb 

(MacOS and Linux)

- Clone and install the feature branch : ok
- Start the client and setup a password : ok
- Create a pmm strategy and exit : ok
- Run the autostart command : ok
- Open a separate terminal and cd to the feature branch : ok
- Run bin/bot_controller_client_demo.py : ok
- Stop the strategy from the controller side : ok
- Start the strategy from the controller side : ok
- Stop and exit using the controller : ok
- Restart the client and run autostart with controller flag : ok
- Run bin/bot_controller_client_demo.py : ok
- Stop the strategy from the controller side : ok
- Start the strategy from the controller side : ok
- Stop and exit using the controller : ok

on going testing

**Comment by gerald-hb** [comment]
Hitting this issue when creating a docker image
image

Steps to reproduce
1.Clone and install feature branch
2.export TAG=name
3.docker build -t coinalpha/hummingbot:name -f Dockerfile .
4.after a few minutes of loading, it will comes up with the error above

**Comment by Deepali6997** [comment]
same error is seen for MacOS.

Screenshot 2022-05-28 at 12 20 29 AM

**Comment by PtrckM** [comment]
- same issue on aarch64 / arm (building docker image)

image

**Comment by petioptrv** [comment]
@gerald-hb, @Deepali6997, @PtrckM, the issue you've flagged also occurs on `development` branch. I tracked it down to a [deprecation](https://docs.conda.io/projects/conda/en/latest/release-notes.html#id180) in one of the flags the docker image passes to `conda clean`. I will open a bug fix for it in another PR and link here.

**Comment by petioptrv** [comment]
@CoinAlpha/qa-team, I've pushed changes to enable using the RPC controller on a docker container, however currently, this is only possible on Linux (linux being the host that starts the docker container)

**Comment by PtrckM** [comment]
- confirmed issue on build client docker image has been fixed. ✅ 

---

- issue when starting the docker instance with `-c` flag or `--controller` flag using [docker strategy autostart](https://hummingbot.org/global-configs/strategy-autostart/#docker-autostart)

image

**Comment by Deepali6997** [comment]
Getting same error on latest commit as flagged by Patrick
Screenshot 2022-05-31 at 2 49 34 PM
steps to reproduce
1. Clone feature branch
2. cd to branch folder
3. export TAG=name
4. docker build -t coinalpha/hummingbot:name -f Dockerfile.arm .
5. Creating PMM strategy
6. Following steps as mentioned in [docker auto start ](https://hummingbot.org/global-configs/strategy-autostart/#how-to-autostart)
7.  -e STRATEGY -e CONFIG_FILE_NAME -e CONFIG_PASSWORD -c\

**Comment by Deepali6997** [comment]
No longer facing the issue as flagged previously
Used only -e STRATEGY -e CONFIG_FILE_NAME -e CONFIG_PASSWORD \

Screenshot 2022-05-31 at 9 17 38 PM

Screenshot 2022-05-31 at 9 26 12 PM

**Review by PtrckM — APPROVED** [review]
LGTM!!!

Test Summary:

- Clone and install the feature branch no issue
  - Start the client and setup a password no issue
  - Create a pmm strategy, start and exit
  - Run the autostart command, confirmed working as expected
- Open a separate terminal and cd to the feature branch then run `bin/bot_controller_client_demo.py` no issue
  - Stop the strategy from the controller side expected output confirmed
  - Start the strategy from the controller side expected output confirmed
  - Stop and exit using the controller expected output confirmed
- Restart the client and run autostart with controller flag then run `bin/bot_controller_client_demo.py` no issue 
  - Stop the strategy from the controller side confirmed working
  - Start the strategy from the controller side confirmed working
  - Stop and exit using the controller confirmed working
- repeat steps and confirmed working using macos (intel/m1), x86 ubuntu and aarch64
- build docker image and confirmed no issue
  - Create a pmm strategy, start and exit
  - Run the autostart command, confirmed working as expected
  - repeat steps on building docker image and confirmed working using macos (intel/m1), x86 ubuntu and aarch64

File: Dockerfile [hunk 1]
```diff
@@ -131,4 +131,4 @@ COPY docker/etc /etc
 # Setting bash as default shell because we have .bashrc with customized PATH (setting SHELL affects RUN, CMD and ENTRYPOINT, but not manual commands e.g. `docker run image COMMAND`!)
 SHELL [ "/bin/bash", "-lc" ]
 CMD /home/hummingbot/miniconda3/envs/$(head -1 setup/environment-linux.yml | cut -d' ' -f2)/bin/python3 bin/hummingbot_quickstart.py \
-    --auto-set-permissions $(id -u hummingbot):$(id -g hummingbot)
+    --controller --auto-set-permissions $(id -u hummingbot):$(id -g hummingbot)
```
[the remark this answers is not in the document]
Comment (petioptrv) [reply Dockerfile:134]: As discussed off of GitHub, I've removed this option and instead added it to the global config — the change is in the PR to `hummingbot-professional`.

File: Makefile [hunk 2]
```diff
@@ -18,3 +22,6 @@ report_coverage:
 development-diff-cover:
 	coverage xml
 	diff-cover --compare-branch=origin/development coverage.xml
+
+rpc-protos:
+	python -m grpc_tools.protoc -I$(RPC_PROTO_DIR) --python_out=$(RPC_PROTO_OUT) --grpc_python_out=$(RPC_PROTO_OUT) $(RPC_PROTO_TARGET)
```
[the remark this answers is not in the document]
Comment (petioptrv) [reply Makefile:27]: It is required. I have created a [task](https://app.shortcut.com/coinalpha/story/28524/add-a-ci-step-to-detect-if-rpc-files-were-modified) to ensure that this gets checked in CI.

File: hummingbot/client/controller/hummingbot_controller.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+        self._app = app
+
+    async def start_command(self):
+        await self._app.start_check()
+
+    async def stop_command(self):
+        await self._app.stop_loop()
+
+    async def exit_command(self):
+        await self._app.exit_loop()
```
[the remark this answers is not in the document]
Comment (petioptrv) [reply hummingbot/client/controller/hummingbot_controller.py:22]: It's possible, but my reasoning for using the async methods is to stay closer to the execution of the commands — the synchronous commands simply schedule the async ones on the event loop and forget about it. By setting this precedent, we can more easily later provide control for things like settings configuration, which may some require back-and-forth.

**Comment by petioptrv** [comment]
Closing this PR in favour of re-opening it towards `hummingbot-professional`: https://github.com/CoinAlpha/hummingbot-professional/pull/22

**Pull request closed without merging**